### PR TITLE
Sending fragments on sync writers <master> [6955]

### PR DESCRIPTION
**Pull request by IkerLuengo**
This is a port of #898 and #902 from 1.9.x

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1452/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1225/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1450/)

File: test/blackbox/BlackboxTestsPubSubFragments.cpp [hunk 1]
```diff
@@ -58,35 +58,65 @@ TEST_P(PubSubFragments, PubSubAsNonReliableData300kb)
     // Mutes an expected error
     Log::SetErrorStringFilter(std::regex("^((?!Big data).)*$"));
```
Comment (MiguelCompany) [inline test/blackbox/BlackboxTestsPubSubFragments.cpp]: These lines are not necessary anymore, as no error should be shown now.

File: test/blackbox/BlackboxTestsPubSubFragments.cpp [hunk 2]
```diff
@@ -58,35 +58,65 @@ TEST_P(PubSubFragments, PubSubAsNonReliableData300kb)
...
-    ASSERT_FALSE(data.empty());
+    ASSERT_TRUE(data.empty());
+    // Block reader until reception finished or timeout.
+    reader.block_for_at_least(2);
 }
 
 TEST_P(PubSubFragments, PubSubAsReliableData300kb)
 {
     // Mutes an expected error
     Log::SetErrorStringFilter(std::regex("^((?!Big data).)*$"));
```
Comment (MiguelCompany) [inline test/blackbox/BlackboxTestsPubSubFragments.cpp]: These lines are not necessary anymore, as no error should be shown now.

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 3]
```diff
@@ -238,9 +238,40 @@ void StatefulWriter::unsent_change_added_to_history(
...
                         RTPSMessageGroup group(mp_RTPSParticipant, this, *this, max_blocking_time);
-                        if (!group.add_data(*change, expectsInlineQos))
+
+                        if (change->getFragmentCount() > 0)
+                        {
+                            ChangeForReader_t change_for_reader(change);
+                            change_for_reader.getUnsentFragments().for_each([&](FragmentNumber_t fragment_number)
+                            {
+                                if (group.add_data_frag(*change, fragment_number,
+                                        expectsInlineQos))
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: This can go in the previous line

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 4]
```diff
@@ -238,9 +238,40 @@ void StatefulWriter::unsent_change_added_to_history(
...
+                                {
+                                    for (ReaderProxy* it : matched_readers_)
+                                    {
+                                        if (!it->is_local_reader())
+                                        {
+                                            bool allFragmentsSent = false;
+                                            it->mark_fragment_as_sent_for_change(
+                                                        change->sequenceNumber,
+                                                        fragment_number,
+                                                        allFragmentsSent);
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: Wrong indentation here

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 5]
```diff
@@ -274,9 +305,26 @@ void StatefulWriter::unsent_change_added_to_history(
...
-                            if (!group.add_data(*change, it->expects_inline_qos()))
+
+                            if (change->getFragmentCount() > 0)
                             {
-                                logError(RTPS_WRITER, "Error sending change " << change->sequenceNumber);
+                                ChangeForReader_t change_for_reader(change);
+                                change_for_reader.getUnsentFragments().for_each([&](FragmentNumber_t fragment_number)
+                                {
+                                    if (!group.add_data_frag(*change, fragment_number,
+                                            it->expects_inline_qos()))
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: This fits in the previous line

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 6]
```diff
@@ -274,9 +305,26 @@ void StatefulWriter::unsent_change_added_to_history(
...
+                                    {
+                                        logError(RTPS_WRITER, "Error sending fragment (" << change->sequenceNumber <<
+                                            ", " << fragment_number << ")");
+                                    }
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: Should we add an `else` here and mark the fragment as sent, the same way it is done on line 255?

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1456/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1229/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1454/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1461/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1234/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1459/)

**Pull request merged**